R/R/prophet.R

```diff
@@ -1274,37 +1268,203 @@ fit.prophet <- function(m, df, ...) {
+    args <- list(
+      object = model,
+      data = dat,
+      init = stan_init,
+      iter = 0,
+      as_vector = FALSE,
+      algorithm = 'Newton',
+      verbose = FALSE
+    )
+    args <- utils::modifyList(args, list(...))
+    stan.fit <- do.call(rstan::optimizing, args)
+    
+    # Extract overall and trend predictions
+    stan.predictions <- 
+      stan.fit$par[c("y_hat", "trend_hat")] %>%
+      as.data.frame() %>%
+      mutate(yhat = y_hat, trend = trend_hat)
+    
+    # Extract overall and trend intervals
+    stan.intervals <- predict_uncertainty(m, stan.fit)
   }
-  df <- df[cols]
-  df <- dplyr::bind_cols(df, seasonal.components, intervals)
-  df$yhat <- df$trend * (1 + df$multiplicative_terms) + df$additive_terms
+  
+  df <- 
+    cbind(
+      ds = df$ds, 
+      stan.predictions, 
+      stan.intervals
+    ) %>% 
+    select( # Retain only overall and trend predictions
+      ds, 
+      yhat, yhat_lower, yhat_upper, 
+      trend, trend_lower, trend_upper
+    ) %>% 
+    mutate_at( # Rescale overall and trend predictions and intervals
+      c(
+        "yhat", "yhat_lower", "yhat_upper",
+        "trend", "trend_lower", "trend_upper"
+      ),
+      funs(. * m$y.scale + df$floor)
+    ) %>% # Now add seasonal breakdowns and intervals
+    cbind(
+      predict_seasonal_components(m, df)
+    )
+  
   return(df)
 }
 
+
+#' Output a list of all data required for stan fitting or prediction
+#'
+#' @param m The fitted prophet object
+#' @param df Data frame with historical or future data for fitting or prediction resp.
+#' @param fit.seasonal.features matrix of seasonal, holiday & exogenous regressors for fitting
+#' @param predict.seasonal.features  matrix of seasonal, holiday & exogenous regressors for predictions
+#' @param component.cols indicators for additive or multiplicative components in seasonal.features
+#' @param prior.scales prior scale hyperparameters
+#' @param type 'fit' or 'predict'
+#' 
+#' @return List with all data required for stan fitting or prediction
+#'
+#' @keywords internal
+data_for_stan <- function(
+  m, df, 
+  fit.seasonal.features, 
+  predict.seasonal.features = matrix(0, 0, ncol(fit.seasonal.features)),
+  component.cols, 
+  prior.scales, 
+  type) {
+  
+  if (type == "fit") {
+    
+    dat <- list(
+      T = nrow(df),
+      K = ncol(fit.seasonal.features),
+      S = length(m$changepoints.t),
+      y = df$y_scaled,
+      t = df$t,
+      cap = switch(
+        m$growth, 
+        linear = rep(0, nrow(df)),                     # Unused inside Stan
+        df$cap_scaled                                  # Add capacities to the Stan data
+      ), 
+      t_change = array(m$changepoints.t),
+      X = as.matrix(fit.seasonal.features),
+      sigmas = array(prior.scales),
+      tau = m$changepoint.prior.scale,
+      trend_indicator = as.numeric(m$growth == 'logistic'),
+      s_a = array(component.cols$additive_terms),
+      s_m = array(component.cols$multiplicative_terms),
```

Comment:
everything from L1397 through 1413 here is repeated below for prediction. 

How about if we construct a list with these regardless of `type`, and then have the `if type == "fit"...` just to add in `T_pred`, `t_pred`, `cap_pred`, `X_pred`, `n_samp`, and `S_pred`?